src/ripple/core/JobCounter.h

```diff
@@ -34,21 +35,33 @@ namespace ripple {
 class JobCounter
 {
 private:
-    std::mutex mutable mutex_ {};
+    std::mutex mutex_ {};
     std::condition_variable allJobsDoneCond_ {};  // guard with mutex_
     bool waitForJobs_ {false};                    // guard with mutex_
-    int jobCount_ {0};                            // guard with mutex_
+    std::atomic<int> jobCount_ {0};
 
-    // Manage the count.  If we're stopping and the count drops to zero
+    // Increment the count.
+    JobCounter& operator++()
+    {
+        ++jobCount_;
+        return *this;
+    }
+
+    // Decrement the count.  If we're stopping and the count drops to zero
     // notify allJobsDoneCond_.
-    void adjustCount (int upDn)
+    JobCounter&  operator--()
     {
+        // Even though jobCount_ is atomic, we decrement its value under a
+        // lock.  This removes a small timing window that occurs if the
+        // waiting thread is handling a spurious wakeup when jobCount_
+        // drops to zero.
         std::lock_guard<std::mutex> lock {mutex_};
 
-        // Count the job.  Notify if we're stopping and all jobs are done.
-        jobCount_ += upDn > 0 ? 1 : -1;
-        if (waitForJobs_ && jobCount_ == 0)
+        // Update jobCount_.  Notify if we're stopping and all jobs are done.
+        // Note that if jobCount_ is 1 pre-decrement then it must become zero.
+        if ((jobCount_.fetch_sub(1) == 1) && waitForJobs_)
```

Comment:
Alternatively, you can use prefix decrement which performs atomically the same (since the default std::memory_order_seq_cst order is used).

`if ((--jobCount_ == 0) && waitForJobs_)`